translationUtils.ts

```diff
@@ -0,0 +1,284 @@
+	if (translations.length === 0 || !isAllSupported) {
+		return null;
+	}
+
+	const translatedImage = await overlayTranslations(image, translations);
+	return translatedImage;
+}
+
+// Returns null if the base64 data could not be retrieved.
+export function getBase64Data(
+	imageUrl: URL
+): Promise<{ base64Data: ImageBase64; width: number; height: number } | null> {
+	return new Promise(resolve => {
+		const image = new Image();
+		image.onload = function () {
+			const canvas = createCanvas(image.width, image.height);
+			const context = canvas.getContext('2d');
+			context.drawImage(
+				image,
+				0,
+				0,
+				image.width,
+				image.height,
+				0,
+				0,
+				image.width,
+				image.height
+			);
+			resolve({
+				base64Data: canvas.toDataURL('image/png'),
+				height: image.height,
+				width: image.width
+			});
+		};
+
+		// Handle various things that can go wrong when loading an image.
+		// This includes network issues, dead links, etc.
+		function onFailed() {
+			resolve(null);
+		}
+		image.onerror = onFailed;
+
+		image.src = imageUrl.toString();
+	});
+}
+
+function checkIsSupportedLanguage(translation: TranslationResult) {
+	const supportedLanguages = ['jp', 'kr', 'cn'];
+	return supportedLanguages.includes(translation.originalLanguage);
+}
+
+function overlayTranslations(
+	imageBase64: string,
+	translations: TranslationResult[]
+): Promise<ImageBase64> {
+	return new Promise(resolve => {
+		const image = new Image();
+		image.onload = function () {
+			const canvas = createCanvas(image.width, image.height);
+			const context = canvas.getContext('2d');
+			context.drawImage(
+				image,
+				0,
+				0,
+				image.width,
+				image.height,
+				0,
+				0,
+				image.width,
+				image.height
+			);
+
+			const filteredTranslations = translations.filter(x => x.translatedText.trim() !== '');
+
+			// Draw text bubble as a rounded rectangle.
+			for (const translation of filteredTranslations) {
+				const width = translation.maxX - translation.minX;
+				const height = translation.maxY - translation.minY;
+				const radius = 10; // Roughly 10px.
+				const backgroundColor = 'white';
+
+				drawRectangle(
+					context,
+					translation.minX,
+					translation.minY,
+					width,
+					height,
+					radius,
+					backgroundColor
+				);
+			}
+
+			// Write text on the bubbles.
+			for (const translation of filteredTranslations) {
+				const maxLineWidth = translation.maxX - translation.minX;
+				const maxHeight = translation.maxY - translation.minY;
+				const startY = translation.minY;
+				const maxFontSizePx = 300; // Arbitrarily large starting font.
+
+				fitText(
+					context,
+					translation.translatedText,
+					translation.minX,
+					startY,
+					maxHeight,
+					maxLineWidth,
+					maxFontSizePx
+				);
+			}
+
+			resolve(canvas.toDataURL('image/png'));
+		};
+
+		image.src = imageBase64;
+	});
+}
+
+// Modified from Grumdrig's StackOverflow response.
```

Comment:
No point in reinventing the wheel :)
I should have included a direct link to the thread, but it's pretty easily Googleable.